### PR TITLE
Issue #3318 - Improve support for Prefer HTTP header

### PR DESCRIPTION
**Pull request by tbieste**
Issue #3318 - Improve support for Prefer HTTP header

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 1]
```diff
@@ -262,43 +264,68 @@ private HTTPHandlingPreference computeHandlingPref(ServletRequest request) throw
...
+
                 if (log.isLoggable(Level.FINE)) {
                     log.fine("Requested return preference = " + returnPref);
                 }
             } catch (IllegalArgumentException e) {
-                String message = "Invalid HTTP return preference passed in header 'Prefer': '" + returnPrefString + "'";
-                if (handlingPref == HTTPHandlingPreference.STRICT) {
-                    throw new FHIRException(message + "; use 'minimal', 'representation' or 'OperationOutcome'.");
+                String message = "Invalid HTTP return preference 'return=" + returnPrefString + "' passed in header 'Prefer'";
+                if (context.getHandlingPreference() == HTTPHandlingPreference.STRICT) {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java:284]: think we need to add to the javadoc to make it clear that the handling preference should be set in the context before calling this method?

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 2]
```diff
@@ -220,39 +218,43 @@ private void checkValidTenantConfiguration(String tenantId) throws Exception {
...
+     * Computes the handling preference from the Prefer header value.
+     * The default handling preference for all interactions is STRICT.
+     * @param context the context containing the HTTP headers
+     * @throws FHIRException
+     */
+    private void computeHandlingPref(FHIRRequestContext context) throws FHIRException {
+        HTTPHandlingPreference handlingPref = HTTPHandlingPreference.from(FHIRConfigHelper.getStringProperty(FHIRConfiguration.PROPERTY_DEFAULT_HANDLING, HTTPHandlingPreference.STRICT.value()));
         boolean allowClientHandlingPref = FHIRConfigHelper.getBooleanProperty(FHIRConfiguration.PROPERTY_ALLOW_CLIENT_HANDLING_PREF, true);
         if (allowClientHandlingPref) {
-            String handlingPrefString = ((HttpServletRequest) request).getHeader(preferHeaderName + ":" + preferHandlingHeaderSectionName);
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java:241]: can we remove the special handling in our FHIRHttpServletRequestWrapper now that we have the processing logic here instead?

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java (hunk 1 above)
Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java:284]: Updated

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java (hunk 2 above)
Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java:241]: Updated.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**